lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -834,20 +834,161 @@ void TypeLoweringVisitor::visitExpr(SubfieldOp op) {
+  FIRRTLType vectorElementType = inputVector.getType()
+                                     .cast<FIRRTLType>()
+                                     .cast<FVectorType>()
+                                     .getElementType();
+  // If this is a multi-dim vector, then substitution will happen at the leaf
+  // node, when the non-vector subelement is accessed.
+  if (vectorElementType.isa<FVectorType>()) {
+    opsToRemove.push_back(op);
+    return;
+  }
+  // Record all the indices used for a multidim access. Default is a 1D vector.
+  SmallVector<SubaccessOp, 8> allDims(1, op);
+  // This is used to build the loop index vector and check all possible values
+  // of the loop indices and select the appropriate vector element after
+  // flattening. For example, all the constants with which to compare with in
+  // this expression : i == 1 ? a_1 : i == 2? a_2 : i == 3 ? a_3 : a_0. For
+  // multi-dim case, the pair of constants such as, <0,1>, <1,0>, <1,1>, <0,0>.
+  // i==0 && j == 1? a_0_1: i == 1 && j == 1 ? a_1_1 and so on.
+  SmallVector<size_t, 8> indicesRange(1, 1);
+
+  // This loop walks back to get the parent vector input on which a multi-dim
+  // access could be performed. If this is a 1D vector, then the loop is  not
+  // executed.
+  while (auto multiDimOp = inputVector.getDefiningOp<SubaccessOp>()) {
+    // Iterate until the input is itself a SubAccess Op.
+    allDims.push_back(multiDimOp);
+    indicesRange.push_back(0);
+    inputVector = multiDimOp.input();
+  }
+  bool isOutput = inputVector.getType().cast<FIRRTLType>().isa<FlipType>();
+  if (auto blockArg = inputVector.dyn_cast<BlockArgument>()) {
+    auto modOp = inputVector.getParentBlock()->getParentOp();
+    auto direction = (Direction)getModulePortDirections(modOp)
+                         .getValue()[blockArg.getArgNumber()];
+    isOutput = direction == Direction::Output;
+  }
+  SmallVector<std::pair<Value, bool>, 8> loweredVector;
+  // Flatten the vector and bundle type recursively and populate the
+  // loweredVector with individual elements after lowering. For example, a[2] is
+  // lowered to a_0,a_1 and  a: {data, valid}[2] is lowered to a_0_data,
+  // a_0_valid, a_1_data, a_1_valid, in that order and the order is important.
+  getAllBundleLowerings(inputVector, loweredVector);
+  // This records all the mux trees generated after lowering op. One mux tree
+  // for each bundle field element.
+  SmallVector<Value, 8> muxTreeVec(1, loweredVector[0].first);
+  // This records a map of bundle subfield name and the corresponding subfield
+  // operation.
+  DenseMap<Attribute, SubfieldOp> subfieldUsers;
+  BundleType vectorOfBundleType = vectorElementType.dyn_cast<BundleType>();
+
+  if (vectorOfBundleType) {
+    // If the vector element is not a bundle, then only one mux tree generated,
+    // else one mux tree for each bundle element. This loop initializes the mux
+    // tree with the first element of the vector (for each element of the
+    // bundle).
```

Comment:
is it possible to have an array inside the bundle, and a subaccess from it?  Does this all compose correctly to N levels?

Reply:
Arrays inside bundle should work, I will add more lit test cases.